### PR TITLE
Add PreparedGrindstoneEvent

### PR DESCRIPTION
**Pull request by JanTuck**
First pull request, hopefully it doesn't suck ass.

### CODE REVIEW

File: Spigot-Server-Patches/0543-Add-PrepareGrindstoneEvent.patch [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+@@ -10,6 +10,12 @@ import org.bukkit.craftbukkit.inventory.CraftInventoryGrindstone;
+ import org.bukkit.craftbukkit.inventory.CraftInventoryView;
+ import org.bukkit.entity.Player;
+ // CraftBukkit end
++// Paper start
++import com.destroystokyo.paper.event.inventory.PrepareGrindstoneEvent;
++import org.bukkit.Bukkit;
++import org.bukkit.craftbukkit.inventory.CraftItemStack;
++// Paper end
++
```
Comment (Proximyst) [inline Spigot-Server-Patches/0543-Add-PrepareGrindstoneEvent.patch]: Whitespace :angry:

File: Spigot-Server-Patches/0543-Add-PrepareGrindstoneEvent.patch [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+ 
+-            this.resultInventory.setItem(0, this.a(itemstack2, i, b0));
++
++            event.setResult(CraftItemStack.asBukkitCopy(this.a(itemstack2, i, b0))); // Paper - Get result
+         } else {
+-            this.resultInventory.setItem(0, ItemStack.a);
++            event.setResult(CraftItemStack.asBukkitCopy(ItemStack.a)); // Paper - Get result
+         }
+ 
++        Bukkit.getPluginManager().callEvent(event); // Paper Event Call
```
Comment (Proximyst) [inline Spigot-Server-Patches/0543-Add-PrepareGrindstoneEvent.patch]: Use https://papermc.io/javadocs/paper/1.15/org/bukkit/event/Event.html#callEvent--

**Review by Proximyst — APPROVED** [review]
All but these two rather minor things look great

File: Spigot-Server-Patches/0543-Add-PrepareGrindstoneEvent.patch (hunk 2 above)
Comment (JanTuck) [reply Spigot-Server-Patches/0543-Add-PrepareGrindstoneEvent.patch]: Will fix.

File: Spigot-Server-Patches/0543-Add-PrepareGrindstoneEvent.patch [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+         ItemStack itemstack1 = this.craftInventory.getItem(1);
+         boolean flag = !itemstack.isEmpty() || !itemstack1.isEmpty();
+         boolean flag1 = !itemstack.isEmpty() && !itemstack1.isEmpty();
++        PrepareGrindstoneEvent event = new PrepareGrindstoneEvent(this.getBukkitView(), null); // Paper
+ 
+         if (flag) {
+             boolean flag2 = !itemstack.isEmpty() && itemstack.getItem() != Items.ENCHANTED_BOOK && !itemstack.hasEnchantments() || !itemstack1.isEmpty() && itemstack1.getItem() != Items.ENCHANTED_BOOK && !itemstack1.hasEnchantments();
+ 
+             if (itemstack.getCount() > 1 || itemstack1.getCount() > 1 || !flag1 && flag2) {
+-                this.resultInventory.setItem(0, ItemStack.a);
```
Comment (aikar) [inline Spigot-Server-Patches/0543-Add-PrepareGrindstoneEvent.patch]: There are many places like this where the event result is set but the function returns early. I don't think this is firing the event here due to the early return.

Comment (JanTuck) [reply Spigot-Server-Patches/0543-Add-PrepareGrindstoneEvent.patch]: I see, that was an oversight on my part

**Pull request closed without merging**